src/byte_set.rs

```diff
@@ -126,21 +132,64 @@ impl<const LEN: usize> From<[u8; LEN]> for ByteSet {
     }
 }
 
-impl From<Range<u8>> for ByteSet {
-    fn from(range: Range<u8>) -> ByteSet {
+impl<'a> From<&'a [u8]> for ByteSet {
+    fn from(bytes: &'a [u8]) -> ByteSet {
         let mut bs = ByteSet::new();
-        for b in range {
-            bs.insert(b);
+        for b in bytes {
+            bs.insert(*b);
         }
         bs
     }
 }
 
-impl From<RangeInclusive<u8>> for ByteSet {
-    fn from(range: RangeInclusive<u8>) -> ByteSet {
-        let mut bs = ByteSet::new();
-        for b in range {
-            bs.insert(b);
+impl From<ops::Range<u8>> for ByteSet {
+    fn from(value: ops::Range<u8>) -> Self {
+        if value.end <= value.start {
+            ByteSet::empty()
+        } else {
+            ByteSet::from(value.start..=value.end - 1)
+        }
+    }
+}
+
+impl From<ops::RangeInclusive<u8>> for ByteSet {
+    fn from(value: ops::RangeInclusive<u8>) -> Self {
+        // because the values are adjacent, we can optimize if they are within the same quadrant
+        let lo = value.start();
+        let hi = value.end();
+        if hi < lo {
+            return ByteSet::empty();
+        } else if hi == lo {
+            return ByteSet::from([*lo]);
+        }
+        let start_ix = *lo / 64;
```

Comment:
Would be tempted to use `>> 6` here instead of division.

Reply:
Done.